airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py

```diff
@@ -731,24 +733,42 @@ def __init__(self, replication_key: str, stream_slice_step: str = "P30D", **kwar
         super().__init__(**kwargs)
         self.replication_key = replication_key
         self._stream_slice_step = stream_slice_step
+        self._cursor = None
+
+    @property
+    def cursor(self) -> Optional[Cursor]:
+        if not self._cursor:
+            raise ValueError("Cursor should be set at this point")
+        return self._cursor
+
+    @cursor.setter
+    def cursor(self, cursor: Cursor):
+        self._cursor = cursor
 
     def stream_slices(
         self, *, sync_mode: SyncMode, cursor_field: List[str] = None, stream_state: Mapping[str, Any] = None
     ) -> Iterable[Optional[Mapping[str, Any]]]:
-        now = pendulum.now(tz="UTC")
-        assert LOOKBACK_SECONDS is not None and LOOKBACK_SECONDS >= 0
-
-        initial_date = self.get_start_date_from_state(stream_state) - pendulum.Duration(seconds=LOOKBACK_SECONDS)
-        slice_start = initial_date
-        while slice_start < now:
-            slice_end = slice_start + self.stream_slice_step
-            self._slice = {
-                "start_date": slice_start.isoformat(timespec="milliseconds"),
-                "end_date": min(slice_end, now).isoformat(timespec="milliseconds"),
-            }
-            yield self._slice
-
-            slice_start += self.stream_slice_step
+        if self._cursor:
+            for slice_start, slice_end in self._cursor.generate_slices():
+                yield {
+                    "start_date": slice_start.isoformat(timespec="milliseconds"),
+                    "end_date": slice_end.isoformat(timespec="milliseconds"),
+                }
+        else:
```

Comment:
shouldn't there always be a cursor? I'd expect either a `FinalStateCursor` or a `ConcurrentCursor`

Reply:
I see. It's because `FinalStateCursor` doesn't create stream slices. I'm not sure it'll get clean up for free with the RFR, but we should make sure to pay back the tech debt. This is getting quite complicated

Reply:
edit: I actually don't get it. creating time-based slices when there's no cursor

Reply:
Partition generation still solely rely on the stream, not on the cursor: https://github.com/airbytehq/airbyte/blob/fad0737d4e3d1f52defd6c3d154f3f2bd85398b9/airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py#L294-L326